pasha/context.py

```diff
@@ -0,0 +1,278 @@
+        context, i.e. with one dimension more than specified by the
+        shape parameter. These are useful for parallelized reduction
+        operations, where each worker may work with its own accumulator.
+
+        Args:
+            Same as array()
+
+        Returns:
+            (numpy.ndarray) Created array object.
+        """
+
+        if isinstance(shape, int):
+            return self.array((self.num_workers, shape), dtype)
+        else:
+            return self.array((self.num_workers,) + tuple(shape), dtype)
+
+    def map(self, function, functor):
+        """Apply a function to a functor.
+
+        This method performs the map operation, applying the function to
+        each element of the functor. The functor may either be an
+        explicit Functor object or any other supported type, which can
+        be wrapped into a default functor.
+
+        Args:
+            function (Callable): Kernel function to map with.
+            functor (Functor or Any): Functor to map over or any type
+                with automatic wrapping support.
+
+        Returns:
+            None
+        """
+
+        raise NotImplementedError('map')
+
+    @staticmethod
+    def run_worker(function, functor, share, worker_id):
+        """Main worker loop.
+
+        This staticmethod contains the actual inner loop for a worker,
+        i.e. iterating over the functor and calling the kernel function.
+
+        Subtypes may call this method after sorting out the required
+        parameters through their specific machinery.
+
+        Args:
+            function (Callable): Kernel function to map with.
+            functor (Functor): Functor to map over.
+            share (Any): Functor's share assigned to this worker.
+            worker_id (int): Identification of this worker. All passed
+                values must be between 0 and num_workers-1.
+
+        Returns:
+            None
+        """
+
+        for entry in functor.iterate(share):
+            function(worker_id, *entry)
+
+
+class SerialContext(MapContext):
+    """Serial map context.
+
+    Runs the map operation directly in the same process and thread
+    without any actual parallelism.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(num_workers=1)
+
+    def map(self, function, functor):
+        functor = Functor.try_wrap(functor)
+        self.run_worker(function, functor, next(iter(functor.split(1))), 0)
+
+
+class PoolContext(MapContext):
+    """Abstract map context for multiprocessing.Pool interface.
+
+    This class contains the common machinery required for a map context
+    based on the Pool interface. A subtype is still required to
+    implemenent the map() method with its actual call signature and then
+    call this type's map method in turn.
+    """
+
+    def __init__(self, num_workers=None):
+        if num_workers is None:
+            num_workers = min(cpu_count() // 2, 10)
```

Comment:
Why not use `cpu_count` by default? I think that's the default of process pool.

Reply:
Hu, I'm not sure which of these two things are the actual reason, so I'll give you both:

1) When doing mostly I/O in your kernels (or at least reading files is still the dominant contribution), I've noticed GPFS to have a sweet spot at 10-20 processes with the performance actually decreasing slightly by adding more processes on the beefy nodes.
2) I've got a deep hatred of people abusing login nodes, so I'd like to do my part of preventing it.

Now 1) of course is a specific (yet historical) application of this stuff, so it shouldn't really be in generic code. But I like to stay a bit more conservative with default values. When someone has a really chunky node and knows what he's doing (provided it's made explicit in the documentation), it's easy to increase it, e.g.

```python
psh.set_default_context('processes', num_workers=999)
```

Since this could be a fairly regular thing, one could make the type even optional (just use the same class).